### PR TITLE
OAuth cleanup & anonymous API access

### PR DESCRIPTION
**Pull request by mworrell**
### Description

Merging https://github.com/zotonic/zotonic/pull/1635 into master.

Also contains:

 * Fixes for OAuth access
 * Force only a single z_datamodel installer running
 * Fix for task queue insert race condition
 * Merge `testsandbox` and `testsandboxdb` sites into one.
 * Limit number of retries for crashed task queue tasks

TODO:

 - [x] Fix OAuth signature calculation (for parameters and binaryfication)
 - [x] Add tests
 - [x] Check templates / functionality

### Checklist

- [x] documentation updated
- [x] tests added
- [x] no BC breaks

### CODE REVIEW

**Comment by mworrell** [comment]
@ddeboer ready for merging!

File: modules/mod_oauth/controllers/controller_oauth_apps.erl [hunk 1]
```diff
@@ -33,29 +33,42 @@ is_authorized(Context) ->
...
-
-
-
-del_consumer(Id, Context) ->
-    m_oauth_app:delete_consumer(Id, Context),
-    Html = z_template:render(<<"_oauth_apps_list.tpl">>, [], Context),
-    Context1 = z_render:update(<<"oauth-apps">>, Html, Context),
-    z_render:wire({growl, [{text, ?__("Application removed.", Context)}]}, Context1).
-
+    case z_convert:to_integer(z_context:get_q("app_id", Context)) of
```
Comment (ddeboer) [inline modules/mod_oauth/controllers/controller_oauth_apps.erl]: `get_q(<<"app_id">>)`: I think we want to standardise on getting binary keys.

File: modules/mod_oauth/controllers/controller_oauth_apps.erl [hunk 2]
```diff
@@ -33,29 +33,42 @@ is_authorized(Context) ->
...
 %% Start add oauth application (consumer)
 %%
-event(#postback{message=start_add_app}, Context) ->
-    z_render:dialog(?__("Add application", Context), <<"_oauth_consumer_edit.tpl">>, [], Context);
+event(Event, Context) ->
+    case z_acl:is_allowed(use, mod_oauth, Context) of
+        true ->
+            do_event(Event, Context);
+        false ->
+            lager:error("Refused access to oauth events."),
```
Comment (ddeboer) [inline modules/mod_oauth/controllers/controller_oauth_apps.erl]: Maybe log the event that was denied?

File: modules/mod_oauth/controllers/controller_oauth_apps.erl [hunk 3]
```diff
@@ -64,48 +77,65 @@ event(#postback{message={start_edit_app, Arg}}, Context) ->
...
-    URL = z_context:get_q(<<"zp-url">>, Context),
-    Callback = z_context:get_q(<<"zp-callback">>, Context),
-    Perms = z_context:get_q_all(<<"zp-perm">>, Context),
-
+do_event(#submit{message={consumer_save, Arg}}, Context) ->
+    Title = z_context:get_q("zp-title", Context),
+    Descr = z_context:get_q("zp-text", Context),
+    URL = z_context:get_q("zp-url", Context),
+    Callback = z_context:get_q("zp-callback", Context),
+    Perms = z_context:get_q_all("zp-perm", Context),
```
Comment (ddeboer) [inline modules/mod_oauth/controllers/controller_oauth_apps.erl]: Again, should the keys be lists or binaries instead?

File: doc/ref/modules/mod_oauth.rst [hunk 4]
```diff
@@ -27,10 +29,35 @@ that particular application.
...
+
+The client will be granted access, though no user will be logged in for the request.
+Because of that the API call will be handled with the same ACL rights as an *anonymous* visitor.
+
+
+User access
+...........
+
+To give an user access, the user will need to register the consumer key and secret with their
+client and then perform the *OAuth authentication" using the web redirects.
```
Comment (ddeboer) [inline doc/ref/modules/mod_oauth.rst]: `*..."` is wrong and causes the doc build to fail.

File: src/support/z_datamodel.erl [hunk 5]
```diff
@@ -50,12 +50,14 @@ manage(Module, Datamodel, Context) ->
...
-    [manage_edge(Module, Edge, Options, AdminContext) || Edge <- Datamodel#datamodel.edges],
-    [manage_data(Module, Data, AdminContext) || Data <- Datamodel#datamodel.data],
+    jobs:run(manage_module_jobs, fun() ->
+        [manage_category(Module, Cat, Options, AdminContext) || Cat <- Datamodel#datamodel.categories],
+        [manage_predicate(Module, Pred, Options, AdminContext) || Pred <- Datamodel#datamodel.predicates],
+        [manage_resource(Module, R, Options, AdminContext) || R <- Datamodel#datamodel.resources],
+        [manage_medium(Module, Medium, Options, AdminContext) || Medium <- Datamodel#datamodel.media],
+        [manage_edge(Module, Edge, Options, AdminContext) || Edge <- Datamodel#datamodel.edges],
+        [manage_data(Module, Data, AdminContext) || Data <- Datamodel#datamodel.data]
+    end),
```
Comment (ddeboer) [inline src/support/z_datamodel.erl]: Out of curiosity: why exactly was this needed?

Comment (mworrell) [inline src/support/z_datamodel.erl]: I saw some tests running into a situation where multiple *category:renumber* jobs were installed in parallel. Apparently we can have multiple modules being enabled at the same time. This serializes all changes, which prevents all kinds of race conditions.

To be fair, I think we should add a serialization on the hierarchy changes as well.

**Comment by ddeboer** [comment]
Thanks a lot!

**Pull request merged**